### PR TITLE
[Feature] optimize PP overlap

### PR DESCRIPTION
**Pull request by Edenzzzz**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs
- [ ] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 📝 What does this PR do?
1. Largely refurbish and clean up interfaces for clarity and maintainability.
2. Optimize Pipeline Parallel, overlapping communication with computation.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

**Comment by Edenzzzz** [comment]
Before overlapping:
image
img_v3_02b6_eefa4ebc-19a6-4ac2-90d8-e455c19dcd7g

File: colossalai/pipeline/p2p.py [hunk 1]
```diff
@@ -324,15 +346,17 @@ def _send_recv_serialization_object(
         return unpickle_object
 
 
+# received_nan = False
```
Comment (ver217) [inline colossalai/pipeline/p2p.py:349]: remove meaningless comment

File: colossalai/pipeline/schedule/one_f_one_b.py [hunk 2]
```diff
@@ -32,20 +32,23 @@ def __init__(
...
             microbatch_size (Optional[int], optional): Microbatch size. If num_microbatches is provided, this will be ignored. Defaults to None.
+            overlap_p2p: Placeholder for interface consistency
         """
         super().__init__(stage_manager)
         assert (
             num_microbatches is not None or microbatch_size is not None
         ), "Either num_microbatches or microbatch_size should be provided"
 
-        self.comm = PipelineP2PCommunication(stage_manager)
+        self.comm = PipelineP2PCommunication(stage_manager, overlap_p2p=False)
```
Comment (ver217) [inline colossalai/pipeline/schedule/one_f_one_b.py:50]: As this is always false, we'd better don't accept this arg (line 35), cause it's a little confusing.

File: colossalai/booster/plugin/hybrid_parallel_plugin.py [hunk 3]
```diff
@@ -1072,13 +1073,15 @@ def __init__(
...
                     enable_metadata_cache=enable_metadata_cache,
+                    overlap_p2p=overlap_p2p,
                 )
             elif pp_style == "1f1b":
                 self.schedule = OneForwardOneBackwardSchedule(
                     stage_manager=self.stage_manager,
                     num_microbatches=num_microbatches,
                     microbatch_size=microbatch_size,
                     enable_metadata_cache=enable_metadata_cache,
+                    overlap_p2p=overlap_p2p,
```
Comment (ver217) [inline colossalai/booster/plugin/hybrid_parallel_plugin.py:1084]: Remove it, as it does not support overlap

File: colossalai/pipeline/schedule/interleaved_pp.py [hunk 4]
```diff
@@ -381,23 +388,27 @@ def run_forward_only(
...
+                input_obj, fwd_wait_handles = self.send_forward_recv_forward(
                     model_chunk_id_send=model_chunk_id,
                     model_chunk_id_recv=self.get_model_chunk_id(i + 1, is_forward=True),
                     output_tensor=output_obj,
-                    send_prior=self.stage_manager.stage % 2 == 0,
+                    send_first=self.stage_manager.stage % 2 == 0,
                 )
             else:
-                self.send_forward(model_chunk_id, output_obj)
+                fwd_wait_handles = self.send_forward(model_chunk_id, output_obj)
```
Comment (ver217) [inline colossalai/pipeline/schedule/interleaved_pp.py:392]: It seems that this part cannot overlap, do we really need async call here?

File: colossalai/pipeline/schedule/interleaved_pp.py [hunk 5]
```diff
@@ -435,35 +448,44 @@ def run_forward_backward(
...
             input_objs[model_chunk_id].append(input_obj)
             output_objs[model_chunk_id].append(output_obj)
 
-            if last_iteration and num_microbatch_remaining == 0:
-                self.send_forward(model_chunk_id, output_obj)
+            if last_batch and num_microbatch_remaining == 0:
+                fwd_wait_handles = self.send_forward(model_chunk_id, output_obj)
             else:
-                input_obj = self.send_forward_recv_forward(
+                input_obj, fwd_wait_handles = self.send_forward_recv_forward(
```
Comment (ver217) [inline colossalai/pipeline/schedule/interleaved_pp.py:451]: If `two_p2p_batches` is disabled, this cannot overlap, right?

File: examples/language/llama/benchmark.py [hunk 6]
```diff
@@ -58,6 +68,7 @@ def main():
         default="gemini",
         help="Choose which plugin to use",
     )
+    parser.add_argument("--overlap", default=True, type=eval)
```
Comment (ver217) [inline examples/language/llama/benchmark.py:71]: Use action="store_true" and this naming should be clearer.

File: examples/language/llama/benchmark.py [hunk 7]
```diff
@@ -256,37 +277,49 @@ def empty_init():
...
-            for step, batch in enumerate(tqdm(dataloader, desc="Step", disable=not coordinator.is_master())):
-                performance_evaluator.on_step_start(step)
-                outputs = model(**batch)
-                loss = outputs[0]
-                booster.backward(loss, optimizer)
-                optimizer.step()
-                optimizer.zero_grad()
-                performance_evaluator.on_step_end(**batch)
-                prof.step()
+            with torch.autocast("cuda", dtype=torch.bfloat16):
```
Comment (ver217) [inline examples/language/llama/benchmark.py:311]: Don't use autocast. Our plugin implements mixed precision

File: examples/language/llama/benchmark.py [hunk 8]
```diff
@@ -256,37 +277,49 @@ def empty_init():
...
+                for step, batch in enumerate(tqdm(dataloader, desc="Step", disable=not coordinator.is_master())):
+                    performance_evaluator.on_step_start(step)
+                    outputs = model(**batch)
+                    loss = outputs[0]
+                    booster.backward(loss, optimizer)
+
+                    if args.profile:
+                        prof.step()
```
Comment (ver217) [inline examples/language/llama/benchmark.py:319]: move it to line 322

File: examples/language/llama/benchmark.py [hunk 9]
```diff
@@ -256,37 +277,49 @@ def empty_init():
...
+                outputs = booster.execute_pipeline(
                     data_iter,
                     model,
                     criterion=lambda outputs, inputs: outputs[0],
                     optimizer=optimizer,
-                    return_loss=False,
+                    return_loss=True,
                 )
-                optimizer.step()
-                optimizer.zero_grad()
```
Comment (ver217) [inline examples/language/llama/benchmark.py:277]: Why not update optimizer?

File: examples/language/llama/pp.sh
Comment (ver217) [inline examples/language/llama/pp.sh:1]: Add readme or don't push it to repo

**Comment by Edenzzzz** [comment]
More benchmark results:
image

File: examples/language/llama/benchmark.py [hunk 10]
```diff
@@ -77,11 +90,12 @@ def main():
     parser.add_argument("--mbs", type=int, default=1, help="Micro batch size of pipeline parallel")
     parser.add_argument("--zero", type=int, default=0, help="Zero Stage when hybrid plugin is enabled")
     parser.add_argument("--custom-ckpt", action="store_true", help="Customize checkpoint", default=False)
-    parser.add_argument("--profile", action="store_true", help="Enable profiling", default=False)
-    parser.add_argument(
-        "--disable-async-reduce", action="store_true", help="Disable the asynchronous reduce operation", default=False
-    )
-    parser.add_argument("--prefetch_num", type=int, default=0, help="chunk prefetch max number")
```
Comment (ver217) [inline examples/language/llama/benchmark.py:84]: Don't remove them

File: examples/language/llama/benchmark.py [hunk 11]
```diff
@@ -77,11 +90,13 @@ def main():
...
-    parser.add_argument(
-        "--disable-async-reduce", action="store_true", help="Disable the asynchronous reduce operation", default=False
-    )
+
+    parser.add_argument("--pp_style", default="1f1b", choices=["1f1b", "interleaved"])
+    parser.add_argument("--n_chunks", default=1, help="number of model chunks", type=eval)
+    parser.add_argument("--profile", action="store_true", help="Profile the code", default=False)
+    parser.add_argument("--disable-async-reduce", action="store_true", help="Disable the asynchronous reduce operation")
     parser.add_argument("--prefetch_num", type=int, default=0, help="chunk prefetch max number")
+    parser.add_argument("--cache", default=True, type=eval)
```
Comment (ver217) [inline examples/language/llama/benchmark.py:99]: use `action="store_true"`

Comment (ver217) [reply examples/language/llama/benchmark.py:99]: or `action="store_false"`

**Review by ver217 — APPROVED** [review]

**Pull request merged**